### PR TITLE
[MR-1] Authentication, login and sign-up

### PR DESCRIPTION
**Pull request by Andrey-2310**

### CODE REVIEW

File: src/main/kotlin/com/kebab/user/controller/AuthenticationController.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/controller/AuthenticationController.kt]: you can do like `@RequestMapping`

File: src/main/kotlin/com/kebab/user/controller/AuthenticationController.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/controller/AuthenticationController.kt]: use static imports, please`methods = [GET, POST])`

File: src/main/kotlin/com/kebab/user/controller/AuthenticationController.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/controller/AuthenticationController.kt]: Try to use static imports everywhere, where it is possible (it makes code easier to read) (not appliable to exception here)

File: src/main/kotlin/com/kebab/user/controller/UserController.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/controller/UserController.kt:46]: why do you need this prefix?

File: src/main/kotlin/com/kebab/user/model/User.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/model/User.kt:49]: :+1:

File: src/main/kotlin/com/kebab/user/repository/UserRepository.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/repository/UserRepository.kt]: no wildcard imports, please

Comment (vtrusevich) [reply src/main/kotlin/com/kebab/user/repository/UserRepository.kt]: it is very bad practice

You can change idea settings, to disable wildcard

File: src/main/kotlin/com/kebab/user/security/AuthenticationInterceptor.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/security/AuthenticationInterceptor.kt:1]: think that we should move auth functionality to separate library

Leave it to me, will do after merging this pr

File: src/main/kotlin/com/kebab/user/security/AuthenticationInterceptor.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/security/AuthenticationInterceptor.kt]: Suggest following refactoring

`override fun preHandle(request: HttpServletRequest, response: HttpServletResponse?, handler: Any?) =
            JwtUtils.getAuthentication(request).run {
                this?.let { SecurityContextHolder.getContext().authentication = this; true }
            } ?: let { response!!.status = HttpServletResponse.SC_FORBIDDEN; false }`

File: src/main/kotlin/com/kebab/user/security/JwtService.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/security/JwtService.kt]: Why have you created it like object?

Comment (vtrusevich) [reply src/main/kotlin/com/kebab/user/security/JwtService.kt]: And as I see, it is not used anywhere

File: src/main/kotlin/com/kebab/user/security/JwtService.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/security/JwtService.kt]: suggest refactoring

[JwtUtil.txt](https://github.com/KebapCase/users-api/files/1781825/JwtUtil.txt)


and suggest moving it to utils package

File: src/main/kotlin/com/kebab/user/security/LoginFilter.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/security/LoginFilter.kt]: suggest

    @Throws(AuthenticationException::class, IOException::class, ServletException::class)
    override fun attemptAuthentication(
            req: HttpServletRequest, res: HttpServletResponse) =
            with(mapper().readValue(req.inputStream, User::class.java)) {
                authenticationManager.authenticate(UsernamePasswordAuthenticationToken(username, password,
                        emptyList<GrantedAuthority>()))
            }!!
}

File: src/main/kotlin/com/kebab/user/controller/AuthenticationController.kt
Comment (vtrusevich) [reply src/main/kotlin/com/kebab/user/controller/AuthenticationController.kt]: review all PR on that please

File: src/main/kotlin/com/kebab/user/security/service/UserDetailsServiceImpl.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/security/service/UserDetailsServiceImpl.kt]: Import?

Comment (vtrusevich) [reply src/main/kotlin/com/kebab/user/security/service/UserDetailsServiceImpl.kt]: I see, that name `User` is already reserved, we can rename our entity to `Account` or something like this

Comment (vtrusevich) [reply src/main/kotlin/com/kebab/user/security/service/UserDetailsServiceImpl.kt]: General recommendation

Kotlin allows to craete functions without specifying exact return type

If you can not specify it, please do not do it

It will help if refactoring in future

File: src/main/kotlin/com/kebab/user/security/util/JwtUtils.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/security/util/JwtUtils.kt]: utils should be utils, not object (see some comments above, there is a sample)

Try to refactor this code, using `?:` `?` `!!` operators and `run` `also` `let` `apply` blocks, please (also samples are above)

File: src/main/kotlin/com/kebab/user/service/UserService.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/service/UserService.kt]: why do you need transactional here?

File: src/main/kotlin/com/kebab/user/service/UserService.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/service/UserService.kt]: Return types!

File: src/main/kotlin/com/kebab/user/service/dto/TokenDto.kt
Comment (vtrusevich) [inline src/main/kotlin/com/kebab/user/service/dto/TokenDto.kt]: I do not see the place, where it is used

File: src/test/kotlin/JwtUtilsTest.kt
Comment (vtrusevich) [inline src/test/kotlin/JwtUtilsTest.kt]: do not declare bracets to the metod (if it is possible), use just `=`

File: src/main/resources/application.properties
Comment (vtrusevich) [inline src/main/resources/application.properties]: you can remove this line

**Review by vtrusevich — CHANGES_REQUESTED** [review]
fix or comment out my comments and suggestions, please

File: src/main/kotlin/com/kebab/user/service/dto/TokenDto.kt
Comment (Andrey-2310) [reply src/main/kotlin/com/kebab/user/service/dto/TokenDto.kt]: AuthenticationController

File: src/main/kotlin/com/kebab/user/security/AuthenticationInterceptor.kt
Comment (Andrey-2310) [reply src/main/kotlin/com/kebab/user/security/AuthenticationInterceptor.kt:1]: Yeah, I also thought about that

**Pull request merged**
